browser/permissions/google_sign_in_permission_browsertest.cc

```diff
@@ -0,0 +1,556 @@
+      permissions::PermissionRequestManager::DISMISS);
+
+  // Have website issue request for Google auth URL
+  ClickButtonWithId(kAuthButtonHtmlId);
+  prompt_factory()->WaitForPermissionBubble();
+
+  // Confirm that content settings did not change
+  EXPECT_EQ(content_settings()->GetContentSetting(
+                embedding_url_, embedding_url_,
+                ContentSettingsType::BRAVE_GOOGLE_SIGN_IN),
+            ContentSetting::CONTENT_SETTING_ASK);
+  EXPECT_EQ(cookie_settings()->GetCookieSetting(
+                GURL(kAccountsGoogleUrl), embedding_url_, nullptr,
+                content_settings::CookieSettingsBase::QueryReason::kCookies),
+            ContentSetting::CONTENT_SETTING_BLOCK);
+
+  EXPECT_EQ(1, prompt_factory()->show_count());
+}
+
+IN_PROC_BROWSER_TEST_F(GoogleSignInBrowserTest, PrefTurnedOff) {
+  SetGoogleSignInPref(false);
+  EXPECT_EQ(0, prompt_factory()->show_count());
+
+  // Try to set 3p cookies from auth domain, should not work
+  NavigateToPageWithFrame(https_cookie_iframe_url_);
+  ExpectCookiesOnHost(GURL(kAccountsGoogleUrl), "");
+  NavigateFrameTo(google_oauth_cookie_url_);
+  ExpectCookiesOnHost(GURL(kAccountsGoogleUrl), "");
+
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), embedding_url_));
+  content::TestNavigationObserver error_observer(contents());
+
+  ClickButtonWithId(kAuthButtonHtmlId);
+  WaitForLoadStopWithoutSuccessCheck(contents());
+  error_observer.Wait();
+  // Navigation blocked
+  EXPECT_EQ(false, error_observer.last_navigation_succeeded());
+  EXPECT_EQ(net::ERR_BLOCKED_BY_CLIENT, error_observer.last_net_error_code());
+
+  EXPECT_EQ(cookie_settings()->GetCookieSetting(
+                GURL(kAccountsGoogleUrl), embedding_url_, nullptr,
+                content_settings::CookieSettingsBase::QueryReason::kCookies),
+            ContentSetting::CONTENT_SETTING_BLOCK);
+
+  // No prompt shown
+  EXPECT_EQ(0, prompt_factory()->show_count());
+}
+
+IN_PROC_BROWSER_TEST_F(GoogleSignInBrowserTest, JSReloadOnBlockedPage) {
+  SetGoogleSignInPref(true);
+  EXPECT_EQ(0, prompt_factory()->show_count());
+
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), embedding_url_));
+  content::TestNavigationObserver error_observer(contents());
+  prompt_factory()->set_response_type(
+      permissions::PermissionRequestManager::DENY_ALL);
+
+  ClickButtonWithId(kAuthButtonHtmlId);
+  WaitForLoadStopWithoutSuccessCheck(contents());
+  error_observer.Wait();
+  // Navigation blocked
+  EXPECT_EQ(false, error_observer.last_navigation_succeeded());
+  EXPECT_EQ(net::ERR_BLOCKED_BY_CLIENT, error_observer.last_net_error_code());
+
+  // Click reload button on blocked page
+  ClickButtonWithId(kReloadButtonHtmlId);
+  WaitForLoadStopWithoutSuccessCheck(contents());
+  error_observer.Wait();
+
+  // Navigation is still blocked
+  EXPECT_EQ(false, error_observer.last_navigation_succeeded());
+  EXPECT_EQ(net::ERR_BLOCKED_BY_CLIENT, error_observer.last_net_error_code());
+}
```

Comment:
please add tests for incognito profile, make sure the setting doesn't leak between profiles.

Reply:
Isn't that just testing for general Chromium behaviour?

Reply:
Not exactly. You need to have some tests for incognito, because permissions in incognito may be auto-blocked and work slightly different. Just make sure they work as you want them to. As a bonus you may check that the setting doesn't leak.

Reply:
Okay, I added a new test for profile and incognito mode. Let me know what you think. I kept the incognito setting mode for the pref to be INHERIT_IF_LESS_PERMISSIVE

Reply:
Got it! Will work on that

Reply:
Hmm, the BRAVE_GOOGLE_SIGN_IN permission is set correctly in incognito, but I see that the cookie rule is not i.e. if we allow 3p cookies for a website in normal profile, then even though the BRAVE_GOOGLE_SIGN_IN is off, we allow 3p cookies. 

This is because in [BravePrefProvider::OnCookieSettingsChanged](https://github.com/brave/brave-core/blob/master/components/content_settings/core/browser/brave_content_settings_pref_provider.cc#L797), we update the cookie rules for both incognito and normal modes. This sets the cookie rules for both regardless of what mode we're in right now. 

I'm guessing I'll have to update that method to check whether or not we're in Incognito before calling `UpdateCookieRules()` for BRAVE_GOOGLE_SIGN_IN permission  but is this fine for the other cookie rules? @boocmp

Reply:
`ContentSettingsType::COOKIES` type uses `ContentSettingsInfo::INHERIT_IN_INCOGNITO` mode, which means all cookie rules we add (including ones generated from `BRAVE_COOKIES`) are also available in incognito. Our shields mode is also shared between normal/incognito profile.

So it looks like this feature should also use `ContentSettingsInfo::INHERIT_IN_INCOGNITO` mode, because it's impossible to separate "shields" cookie rules from "google sign in" cookie rules.

Reply:
Interesting. Okay, that makes sense.